interface/modules/custom_modules/oe-module-faxsms/rcauth.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+/**
+ * Fax SMS Module Member
+ *
+ * @package   OpenEMR
+ * @link      http://www.open-emr.org
+ * @author    Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2018-2021 Jerry Padgett <sjpadgett@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+$ignoreAuth = 1;
+$sessionAllowWrite = true;
+require_once(__DIR__ . "/../../../globals.php");
+
+use OpenEMR\Common\Crypto\CryptoGen;
+use RingCentral\SDK\SDK;
+
+$url = $_SESSION['url'];
+$callbackUrl = $_SESSION['redirect_uri'];
+
+function processCode(): void
+{
+    $vendor = '_ringcentral';
+    $cacheDir = $GLOBALS['OE_SITE_DIR'] . '/documents/logs_and_misc/_cache';
+    $authUser = 0;
+    $credentials = sqlQuery("SELECT * FROM `module_faxsms_credentials` WHERE `auth_user` = ? AND `vendor` = ?", array($authUser, $vendor))['credentials'];
+    if (empty($credentials)) {
+        echo xlt("Applications credentials were not found. Please setup account.");
+        die('Credential Error');
+    }
+    $cryptoGen = new CryptoGen();
+    $credentials = json_decode($cryptoGen->decryptStandard($credentials), true, 512, JSON_THROW_ON_ERROR);
+    $serverUrl = !$credentials['production'] ? "https://platform.devtest.ringcentral.com" : "https://platform.ringcentral.com";
+
+    $callbackUrl = $credentials['redirect_url'];
+    $rcsdk = new SDK($credentials['appKey'], $credentials['appSecret'], $serverUrl, 'OpenEMR', '1.0.0');
+    $platform = $rcsdk->platform();
+
+    $qs = $platform->parseAuthRedirectUrl($_SERVER['QUERY_STRING']);
+    $qs["redirectUri"] = $callbackUrl;
+
+    // log in
+    $apiResponse = $platform->login($qs);
+    $_SESSION['sessionAccessToken'] = $apiResponse->text();
+
+    // archive authentication data for future reauths.
+    $file = $cacheDir . DIRECTORY_SEPARATOR . 'platform.json';
+    $content = json_encode($platform->auth()->data(), JSON_THROW_ON_ERROR | JSON_PRETTY_PRINT);
+    $cryptoGen = new CryptoGen();
+    $content = $cryptoGen->encryptStandard($content);
```

Comment:
In this case, you are encrypting something to the drive, so you should use the key from the database:
```
$content = $cryptoGen->encryptStandard($content, null, 'database');
```

Reply:
okay thanks.

Reply:
I wonder how it was working! I better check my main encryption in AppDispatch.

Reply:
yep, the theory is that if:
1. The database is compromised (but not the drive), then at least encrypted items are safe
2. The drive is compromised (but not the database), then at least encrypted items are safe
3. If both database and drive are compromised, then all is lost (except for the hashed passwords of course :) )